src/modules/job-exec/bulk-exec.c

```diff
@@ -0,0 +1,400 @@
+        else
+            flux_log (exec->h, LOG_INFO, "rank %d: %s: %s", rank, stream, s);
+    }
+}
+
+static void exec_cmd_destroy (void *arg)
+{
+    struct exec_cmd *cmd = arg;
+    idset_destroy (cmd->ranks);
+    flux_cmd_destroy (cmd->cmd);
+    free (cmd);
+}
+
+static struct exec_cmd *exec_cmd_create (const struct idset *ranks,
+                                         flux_cmd_t *cmd,
+                                         int flags)
+{
+    struct exec_cmd *c = calloc (1, sizeof (*c));
+    if (!c)
+        return NULL;
+    if (!(c->ranks = idset_copy (ranks)))
+        goto err;
+    c->cmd = flux_cmd_copy (cmd);
+    c->flags = flags;
+    return (c);
+err:
+    exec_cmd_destroy (c);
+    return NULL;
+}
+
+static void subprocess_destroy_finish (flux_future_t *f, void *arg)
+{
+    flux_subprocess_t *p = arg;
+    if (flux_future_get (f, NULL) < 0) {
+        flux_t *h = flux_subprocess_aux_get (p, "flux_t");
+        flux_log_error (h, "subprocess_kill: %ju: %s",
+                        (uintmax_t) flux_subprocess_pid,
+                        flux_strerror (errno));
+    }
+    flux_subprocess_destroy (p);
+    flux_future_destroy (f);
+}
+
+static int subprocess_destroy (flux_t *h, flux_subprocess_t *p)
+{
+    flux_future_t *f = flux_subprocess_kill (p, SIGKILL);
+    if (!f || flux_future_then (f, -1., subprocess_destroy_finish, p) < 0)
+        return -1;
+    return 0;
+}
+
+static int exec_start_cmd (struct bulk_exec *exec,
+                           struct exec_cmd *cmd,
+                           int max)
+{
+    int count = 0;
+    uint32_t rank;
+    rank = idset_first (cmd->ranks);
+    while (rank != IDSET_INVALID_ID && (max < 0 || count < max)) {
+        flux_subprocess_t *p = flux_rexec (exec->h,
+                                           rank,
+                                           cmd->flags,
+                                           cmd->cmd,
+                                           &exec->ops);
+        if (!p)
+            return -1;
+        if (flux_subprocess_aux_set (p, "job-exec::exec", exec, NULL) < 0
+           || zlist_append (exec->processes, p) < 0) {
+            if (subprocess_destroy (exec->h, p) < 0)
+                flux_log_error (exec->h, "Unable to destroy pid %ju",
+                        (uintmax_t) flux_subprocess_pid (p));
+            return -1;
+        }
+        zlist_freefn (exec->processes, p,
+                     (zlist_free_fn *) flux_subprocess_unref,
+                     true);
+
+        idset_clear (cmd->ranks, rank);
+        rank = idset_next (cmd->ranks, rank);
+        count++;
+    }
+    if (idset_count (cmd->ranks) == 0)
+        zlist_remove (exec->commands, cmd);
+    return count;
+}
+
+void bulk_exec_stop (struct bulk_exec *exec)
+{
+    flux_watcher_stop (exec->prep);
+    flux_watcher_stop (exec->check);
+}
+
+static int exec_start_cmds (struct bulk_exec *exec, int max)
```

Comment:
noticed `max` is only called with `-1` set and isn't configurable via something in the API.  Perhaps don't need this?  Or later allow the user to set some type of "flow-control"-ish kinda setting?

Reply:
yes, this was for later. Likely not needed for the near future, but I need to get this PR done before I go so unfortunately didn't want to focus on that right now.

Note that also there is placeholder code for partial release which is unused. If it is ok, I'd like that to stay in as well.

Reply:
Actually I ended up making the "max rexec per loop" a configurable parameter in bulk_exec, set to 1 by default. This will prevent the job-exec module loop from "pausing" to call `flux_subprocess_rexec()` for large-ish jobs and more fairly handle other exec module traffic (however, it probably won't be an issue either way)

At least now we could tune it (and I don't have to dump the existing code)